Reimplement snapshots as a new implementation of Engine interface.

spencerkimball opened the pull request:
This fixes a TODO in the code and helpful to upcoming changes to
implement asynchronous scans of ranges for custodial work. Without,
more Engine method variants taking snapshot IDs would be necessary.
This change is more in keeping with how we do Batch engines and is
simpler.

Also remove the InternalSnapshotCopy method, which will be replaced
by an in-Raft mechanism whereby the leader feeds the snapshot directly
to a requesting follower.

Commit:
Reimplement snapshots as a new implementation of Engine interface.

This fixes a TODO in the code and helpful to upcoming changes to
implement asynchronous scans of ranges for custodial work. Without,
more Engine method variants taking snapshot IDs would be necessary.
This change is more in keeping with how we do Batch engines and is
simpler.

Also remove the InternalSnapshotCopy method, which will be replaced
by an in-Raft mechanism whereby the leader feeds the snapshot directly
to a requesting follower.

spencerkimball commented:
@bdarnell

petermattis commented on storage/range_test.go:
```diff
@@ -835,113 +816,6 @@ func TestRangeIdempotence(t *testing.T) {
 }
 
-// TestRangeSnapshot.
-func TestRangeSnapshot(t *testing.T) {
```
I would think you'd want some tests that the various snapshot methods do what you expect. Am I just missing where this testing is being done?

spencerkimball replied:
OK, added unittests for all snapshot methods.

Commit:
Add unittest verifiying all methods of snapshot engines.

Changed NewSnapshot behavior to log an error and return nil if called
from a snapshot.

bdarnell commented on storage/engine/rocksdb.go:
```diff
@@ -428,7 +364,20 @@ func (r *RocksDB) NewIterator() Iterator {
+// read-only rocksDBSnapshot engine.
+func (r *RocksDB) NewSnapshot() Engine {
+	if r.rdb == nil {
+		log.Errorf("RocksDB is not initialized yet")
```
In general I think it's better to panic than to log an error and return nil (when we cannot return an `error`) (here and inMemSnapshot.New{Batch,Snapshot})

spencerkimball replied:
Addressed in #257

bdarnell commented:
LGTM
